---
title: "Circle Plots and Enrichment Analysis"
author: "Rhonda Bacher"
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---


# R packages needed

```{r, cache=TRUE, message=FALSE, results='hide'}
library(circlize)
library(GenomicRanges)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
library(biomaRt)
library(org.Mm.eg.db)
library(ComplexHeatmap)
```


# Load in and format the diabetes SNPs obtained from GWAS Central for circle plot

This will provide the data to make the diabetes SNPs histogram on the human chromosomes on the circle plot.
Counts how many SNPS rewithin a 1Mbp window.

```{r, cache=TRUE, message=FALSE, results='hide'}
# Format the Type 1 and 2 Diabetes GWAS SNPs from GWAS Central:

diab.all <- read.table("human_diabetesSNPs_forSupplement.csv", sep=",", header=T, stringsAsFactors = FALSE)
diab.type1 <- subset(diab.all, trait == "Type 1 Diabetes")
diab.type2 <- subset(diab.all, trait == "Type 2 Diabetes")

diab.type1$value <- 1
diab.type2$value <- 1

cytoband_hg = read.cytoband(species = "hg19")
cytoband_df_hg = cytoband_hg$df
cytoband_df_hg <- subset(cytoband_df_hg, V1!="chrY")

windows <- floor(cytoband_hg$chr.len[-24] / 1000000)
chroms <- names(windows)

myVect <- c()
for(i in 1:length(chroms)) {
  toSUM <- which(diab.type1$chr == chroms[i])
  for(j in 1:windows[i]) {
    Upper <- j * 1000000
    Lower <- (j-1) * 1000000
    wSum <- length(which(diab.type1[toSUM,"start"] <= Upper & diab.type1[toSUM,"start"] > Lower))
    myVect <- c(myVect, wSum)
  }
}
intervalS <- (do.call(c, sapply(windows, function(x) seq(1:x)))) * 1000000
plotType1 <- data.frame(chr = (rep(names(windows), windows)), start = intervalS - 1000000,
                          end=intervalS, value = myVect, stringsAsFactors = FALSE)

myVect <- c()
for(i in 1:length(chroms)) {
  toSUM <- which(diab.type2$chr == chroms[i])
  for(j in 1:windows[i]) {
    Upper <- j * 1000000
    Lower <- (j-1) * 1000000
    wSum <- length(which(diab.type2[toSUM,"start"] <= Upper & diab.type2[toSUM,"start"] > Lower))
    myVect <- c(myVect, wSum)
  }
}
intervalS <- (do.call(c, sapply(windows, function(x) seq(1:x)))) * 1000000
plotType2 <- data.frame(chr = (rep(names(windows), windows)), start = intervalS - 1000000,
                        end=intervalS, value = myVect, stringsAsFactors = FALSE)

# Due to area on chr 6 with lots of SNPs fix maximum in the plot.
plotType1[,4][plotType1[,4] > 20] <- 20
```

List of the hoptspot genes to be included on the plot

```{r, cache=TRUE, message=FALSE, results='hide'}
hotspotGenes<- data.frame(chr=c("chr13", "chr20", "chr19", "chr5", "chr17"),
                       start=c(27919994, 44355700, 48600600, 55935095, 7626234),
                       end = c(27926314, 44434596, 48614854, 55994993, 7627876),
                       value= runif(5, 1.3, 1.31),
                      gene = c("PDX1","HNF4A", "FAM83E", "IL6ST", "SAT2"),
                      stringsAsFactors = F)

```



## Circos plot for clinical QTL

First need to set up the chromosome information for the circle plot.

```{r, cache=TRUE, message=FALSE, results='hide'}

# Set up base of circle plot using tracks in circlize package:
cytoband_hg = read.cytoband(species = "hg19")
cytoband_df_hg = cytoband_hg$df
cytoband_df_hg <- subset(cytoband_df_hg, V1!="chrY")

cytoband_mm = read.cytoband(species = "mm10")
cytoband_df_mm = cytoband_mm$df
cytoband_df_mm <- cytoband_df_mm[nrow(cytoband_df_mm):1,]
cytoband_df_mm <- subset(cytoband_df_mm, V1!="chrY")
cytoband_df_mm$V1 <- paste0("mm10_", cytoband_df_mm$V1)

xrange = c(cytoband_hg$chr.len[1:23], rev(cytoband_mm$chr.len)[-1])
human_chr_index = 1:23
mouse_chr_index = 24:43
names(xrange)[mouse_chr_index] <-  paste0( "mm10_", names(xrange)[mouse_chr_index])

sector.width = c(xrange[human_chr_index] / sum(xrange[human_chr_index]),
                 xrange[mouse_chr_index] / sum(xrange[mouse_chr_index]))

Both_Genome <- rbind(cytoband_df_hg, cytoband_df_mm)
Both_Genome$V1 <- factor(Both_Genome$V1, levels = unique(Both_Genome$V1))
NAMES <- c(1:22, "X","X", 19:1)

```

Read in the clinical QTL data, get nearest gene, and get human ortholog:

```{r, cache=TRUE, message=FALSE, results='hide'}
### Now read in clinical QTL
clin_data <- read.csv("pheno_clin_QTL_peaks.csv", header=T, stringsAsFactors = F)
clin_data <- clin_data[,c(2,3,5,6,4,1)] #reorder
colnames(clin_data)[3:4] <- c("CIStart", "CIEnd")
clin_data[,c(2:4)] <- clin_data[,c(2:4)] * 1000000

# ChIPseeker package has nice function to get nearest gene. To use it for the closest gene
# need a small region
get_NearestGenes <- data.frame(clin_data)
get_NearestGenes$start <- clin_data[,2] - 1
get_NearestGenes$end <- clin_data[,2] + 1
get_NearestGenes <- get_NearestGenes[,c(1,7,8,2:6)]

get_NearestGenes$chr <- paste0("chr", get_NearestGenes$chr) # Need the chr prefix

gData <- makeGRangesFromDataFrame(get_NearestGenes, keep.extra.columns=TRUE,
                            ignore.strand=TRUE,
                            seqnames.field = "chr",
                            start.field="start", end.field="end")
peakAnno <- annotatePeak(gData, tssRegion=c(-1, 1),
                         TxDb=txdb, annoDb="org.Mm.eg.db",
                         ignoreOverlap = FALSE,
                         overlap="all")
toUse <- data.frame(peakAnno@anno)

# Had to manually annotate a few, I just used UCSC genome browser and
# for each QTL position looked for nearest gene that was not a predicted gene or
# gene model because those will not have orthologs.
# These were all the closest to the QTL and within the orginal QTL confidence interval.
toUse <- toUse[,c(1, 6:9, 10, 19, 21)]
toUse[2, "SYMBOL"] <- "Pramel3"
toUse[10, "SYMBOL"] <- "Stard13"
toUse[15, "SYMBOL"] <- "Actrt1"
toUse[18, "SYMBOL"] <- "Actrt1"
toUse[26, "SYMBOL"] <- "Enc1"
toUse[30, "SYMBOL"] <- "Ccl6"

### Download ortholog information from biomaRt:
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
MMattributes = c("ensembl_gene_id","hsapiens_homolog_ensembl_gene", "hsapiens_homolog_perc_id")
orth.mouse = getBM(MMattributes, filters="with_hsapiens_homolog",values =TRUE, mart = mouse, bmHeader=FALSE)

human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
HGattributes <- c("chromosome_name", "start_position", "end_position", "hgnc_symbol","ensembl_gene_id")
human_convert <- getBM(HGattributes, "ensembl_gene_id", orth.mouse$hsapiens_homolog_ensembl_gene, mart= human)
mouse_convert <- getBM(c("mgi_symbol","ensembl_gene_id"), "ensembl_gene_id", orth.mouse$ensembl_gene_id, mart= mouse)

orth.1 <- merge(orth.mouse, mouse_convert, by="ensembl_gene_id", all=TRUE)
orth.2 <- merge(orth.1, human_convert, by.x="hsapiens_homolog_ensembl_gene", by.y="ensembl_gene_id", all.x=TRUE)
# will use orth.2

## Now merge back the ortholog info in.
Hgenes <- orth.2[which(orth.2$mgi_symbol %in% toUse$SYMBOL),c(4:8, 3)]

# Check any with no ortholog?
setdiff(toUse$SYMBOL, Hgenes$mgi_symbol)
subset(toUse, !(toUse$SYMBOL %in%Hgenes$mgi_symbol))

# This one does have one that I found by searching manually,
# Not sure why it didn't show up here. I've added it directly:
Hgenes <- rbind(Hgenes, data.frame(mgi_symbol=setdiff(toUse$SYMBOL, Hgenes$mgi_symbol),
	 chromosome_name=c(22),
	 start_position=c(22890123),
	 end_position=c(22901768),
	 hgnc_symbol=c("PRAME"),
	 hsapiens_homolog_perc_id=c(NA)))

Hgenes[,2] <- paste0("chr", Hgenes[,2])
Hgenes <- Hgenes[,c(2:4,1,5,6)]
names(Hgenes)[1:4] <- c("chr", "start", "end", "SYMBOL")

toUse[,1]<-paste0("mm10_", toUse[,1])

# merge orthologs and clinical qtl together
ORTHO <- merge(Hgenes, toUse, by="SYMBOL", all=F)

# Some mouse genes map to multiple orthologs in human
# Remove duplicates
# Choose based on percent homology or matching symbol name
subset(ORTHO, ORTHO$SYMBOL %in% unique(ORTHO[duplicated(ORTHO[,7:9]),1]))
ORTHO <- ORTHO[-c(7,8,36),]


# The following is formatting for the plot and creating the supplement file:
# We want the names to look nice for the plot:
clin_data[,1] <- paste0("mm10_chr", clin_data[,1])
colnames(clin_data) <- c("chr.mm", "qtl", "qtl.lowCI", "qtl.highCI",  "lod", "lodcolumn")
clin_data$name.for.fig <- c("Glucose at 6 wk", "TG at 6 wk", "Glucose at 10 wk",
"Insulin at 10 wk", "Insulin at 10 wk", "TG at 10 wk", "TG at 10 wk", "TG at 10 wk",
"Insulin at 14 wk","Insulin at 14 wk","Insulin at 14 wk","Insulin at 14 wk",
"TG at 14 wk", "Weight at 2 wk", "Weight at 2 wk", "Weight at 6 wk", "Weight at 6 wk",
"Weight at 6 wk", "Weight at 10 wk", "Weight at 10 wk", "HOMA-B", "HOMA-B", "HOMA-B", "HOMA-B",
"HOMA-IR", "HOMA-IR", "HOMA-IR", "HOMA-IR", "AUCglucose", "AUCinsulin", "Islets/pancreas", "Insulin/islet", "WPIC", "Food consumption")

# Additional manual formatting for the plot
allHMLG_data <- ORTHO[order(ORTHO$seqnames, ORTHO$CIStart),]
h_data <- allHMLG_data[,2:4]
h_data[,2] <- round(as.numeric(h_data[,2]) / 1000000) #nearest Mbp
h_data[,3] <- h_data[,2] + 1
h_data[,2:3] <- h_data[,2:3] * 1000000
colnames(h_data) <- colnames(ORTHO[,2:4])

anno_data <- clin_data[,c(1,3,4)]
colnames(anno_data) <- colnames(ORTHO[,2:4])
anno_data <- rbind(anno_data, h_data)
anno_data$value <- 1
anno_data$value[1:34] <- clin_data[,7]
anno_data$value[35:nrow(anno_data)] <- anno_data[35:nrow(anno_data),2] / 1000000
anno_data$cols1 <- (c(anno_data$value[1:34], rep("black",nrow(anno_data) - 34)))
anno_data <- anno_data[order(factor(anno_data[,1], levels = levels(Both_Genome$V1)), anno_data[,2], anno_data[,3]),]

B_hg <- h_data[,1:3]
B_mm <- allHMLG_data[,c(7,9,10)]

## This is to format the data as a supplementary file for the paper:
toSupp <- allHMLG_data
colnames(toSupp) <- c("mgi_symbol", "chr.hg", "start.hg", "end.hg", "hgnc_symbol", "percent_homology", "chr.mm", "qtl", "qtl.lowCI", "qtl.highCI", "lod", "lodcolumn", "distanceToTSS")
toSupp <- merge(toSupp, clin_data, by=c("chr.mm", "qtl.lowCI", "qtl.highCI", "qtl"))
toSupp[,4] <- round(as.numeric(toSupp[,4]) / 1000000)
toSupp[,7] <- round(as.numeric(toSupp[,7]) / 1000000)
toSupp[,6] <- paste0("hg19_", toSupp[,6])
toSupp[,11] <- round(as.numeric(toSupp[,11]), 2)
#better organizing/names for output
toSupp <- toSupp[,c(16,1,4,11,5,9,6,7)]
colnames(toSupp) <- c( "trait", "mm10_chr", "qtl_Mbp", "lod.score", "nearest_gene_mgi","syntenic_gene_hg","syntenic_hg19_chr", "syntenic_hg19_Mbp")
clinData <- toSupp
# write.table(toSupp, file="tableInfo_clinicalTraits_forSupplement.csv", quote=F, row.names=F, sep=",")

```

Plotting the clinical data circle plot

```{r, fig.height=20, fig.width=20, cache=TRUE, message=FALSE}
set.seed(2020)
colfunc <- colorRampPalette(c("paleturquoise1", "blue"))
COL1 <- "#a8df49" #colfunc(24)[4]
colfunc2 <- colorRampPalette(c("orange", "red"))
COL2 <- colfunc2(21)[2]

circos.clear()
circos.par(start.degree = 85, "gap.degree" = c(rep(1, 22), 10, rep(1, 19), 10),
           canvas.xlim=c(-.1,.1), canvas.ylim=c(-.95, .95), track.margin=c(.01, .01),
           cell.padding=c(.01, .01, .01, .01), track.height = 0.3, points.overflow.warning = FALSE)

circos.initializeWithIdeogram(plotType=NULL, Both_Genome, sector.width = sector.width, 
                              sort.chr = FALSE, track.height=.8)

circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = .2)
circos.genomicTrack(plotType1, track.index=1, bg.border="white", ylim=c(0,20),
                    panel.fun = function(region, value, ...) {
                      circos.genomicLines(region, value, col = "brown3", type='l', lwd=4)
                    })
circos.yaxis(side = "left", sector.index = "chr1", track.index=1, labels.cex=1.1, col="black", labels.col="black")
circos.genomicTrack(plotType2, track.index=1, bg.border="white", ylim=c(0,20),
                    panel.fun = function(region, value, ...) {
                      circos.genomicLines(region, value, col = "dodgerblue", type='l', lwd=3)
                    })
circos.genomicTrack(hotspotGenes, track.index=1, stack=T, bg.border="white",
                    numeric.column = 4, 
                    panel.fun = function(region, value, ...) { 
                      circos.genomicLines(region, value, col = "red", type="h", lty=2)
                    })
circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = .8)
circos.genomicLabels(anno_data, labels.column = 4, cex=1.5, track.margin=c(0,0), 
                     side="outside", padding=.1, col="black", 
                     connection_height=convert_height(10, "mm"), line_lwd=4,
                     labels_height = (convert_height(1, "cm")))
circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = 0.1)

COLORS <- c(rep(COL1, 23), rep(COL2, 20))
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
  chr = CELL_META$sector.index
  xlim = CELL_META$xlim
  ylim = CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col = COLORS[CELL_META$sector.numeric.index], border="white")
  circos.text(mean(xlim), mean(ylim), NAMES[CELL_META$sector.numeric.index], col = "black", cex=1.5,
              facing="downward", niceFacing = TRUE)
}, track.height = 0.25, bg.border = NA)

circos.par("track.height" = 0.1, cell.padding = c(.1, .1, .1, .1))

circos.genomicLink(B_mm, B_hg, col=gray.colors(34, start = .1, end=.8, alpha=1))

```


## Circos plot for module QTL

Read in the module QTL data, get nearest gene, and get human ortholog:
```{r, cache=TRUE, message=FALSE, results='hide'}

### Now read in molecular QTL
mod_data <- read.csv("ME_QTL_peaks.csv", header=T, stringsAsFactors = F)
mod_data <- mod_data[,c(3,6:7, 2, 4, 5)]
colnames(mod_data)[2:3] <- c("Ostart", "Oend")
mod_data[,c(2,3,5)] <- mod_data[,c(2,3,5)] * 1000000

get_NearestGenes <- data.frame(mod_data)

# Want the closest gene the the QTL, need small region
get_NearestGenes$start <- mod_data[,5] - 1
get_NearestGenes$end <- mod_data[,5] + 1
get_NearestGenes <- get_NearestGenes[,c(1,7,8,2:6)]

get_NearestGenes$chr <- paste0("chr", get_NearestGenes$chr) # Need the chr prefix

gData <- makeGRangesFromDataFrame(get_NearestGenes, keep.extra.columns=TRUE, 
                                  ignore.strand=TRUE,
                                  seqnames.field = "chr",
                                  start.field="start", end.field="end")
peakAnno <- annotatePeak(gData, tssRegion=c(-1, 1),
                         TxDb=txdb, annoDb="org.Mm.eg.db", 
                         ignoreOverlap = FALSE, 
                         overlap="all")
toUse <- data.frame(peakAnno@anno)
toUse <- toUse[,c(1, 6:9, 10, 19, 21)]

# Had to manually annotate these below, I just used UCSC genome browser and 
# for each QTL position looked for nearest gene that was not a predicted gene 
# or gene model. These are all the nearest within the orginal QTL confidence interval.
toUse[9, "SYMBOL"] <- "Plod3"
toUse[13, "SYMBOL"] <- "Gdap2"
toUse[15, "SYMBOL"] <- "Hmga2"
toUse[43, "SYMBOL"] <- "Usp27x"
toUse[45, "SYMBOL"] <- "Lamtor4"
toUse[48, "SYMBOL"] <- "Flnb"
toUse[50, "SYMBOL"] <- "Nr3c1"
toUse[52, "SYMBOL"] <- "Vgf"
toUse[56, "SYMBOL"] <- "Nr3c1"
toUse[63, "SYMBOL"] <- "Nlrp1b"
toUse[66, "SYMBOL"] <- "Igsf1"
toUse[77, "SYMBOL"] <- "Myot"

## Now merge back the ortholog info in.
Hgenes <- orth.2[which(orth.2$mgi_symbol %in% toUse$SYMBOL),c(4:8, 3)]
Hgenes <- Hgenes[!duplicated(Hgenes),]
Hgenes <- Hgenes[which(Hgenes$hgnc_symbol != ""),]

# Any with no otherlog? No
subset(toUse, !(toUse$SYMBOL %in%Hgenes$mgi_symbol))

Hgenes[,2]<-paste0("chr", Hgenes[,2])
Hgenes <- Hgenes[,c(2:4,1,5,6)]
names(Hgenes)[1:4] <- c("chr", "start", "end", "SYMBOL")

toUse[,1]<-paste0("mm10_", toUse[,1])

ORTHO <- merge(Hgenes, toUse, by="SYMBOL", all=F)

# Some mouse genes map to multiple orthologs in human
# Choose based on percent homology or matching symbol name
subset(ORTHO, ORTHO$SYMBOL %in% unique(ORTHO[duplicated(ORTHO[,7:10]),1]))
ORTHO <- ORTHO[-c(1,31),]

mod_data[,1] <- paste0("mm10_chr", mod_data[,1])
colnames(mod_data) <- c("chr.mm", "qtl.lowCI", "qtl.highCI",  "lodcolumn", "qtl",  "lod")

## Need to add in enrichment labels for each module...
mod_data$lodcolumn <- gsub("ME", "", mod_data$lodcolumn)
library(readxl)
colorKey <- data.frame(read_excel("Top GO terms for each module.xlsx"), stringsAsFactors=FALSE)[,2:3]
colnames(colorKey) <- c("lodcolumn", "Term")

mod_data <- merge(mod_data, colorKey, by="lodcolumn", all.x=T, all.y=F)
mod_data <- mod_data[!duplicated(mod_data),]


# more formatting for the plot
allHMLG_data <- ORTHO[order(ORTHO$seqnames, ORTHO$Ostart),]
h_data <- allHMLG_data[,2:4]
h_data[,2] <- round(as.numeric(h_data[,2]) / 1000000)
h_data[which(is.na(h_data$start)),2] <- h_data[which(is.na(h_data$start)),3]
h_data[,3] <- h_data[,2] + 1
h_data[,2:3] <- h_data[,2:3] * 1000000
colnames(h_data) <- colnames(ORTHO[,2:4])

anno_data <- mod_data[,c(2,3,4)]
colnames(anno_data) <- colnames(ORTHO[,2:4])
anno_data <- rbind(anno_data, h_data)
anno_data$value <- 1
anno_data$value[1:77] <- mod_data[,1]
anno_data$value[78:nrow(anno_data)] <- anno_data[78:nrow(anno_data),2] / 1000000
anno_data$cols1 <- (c(anno_data$value[1:77], rep("black",nrow(anno_data) - 77)))
anno_data <- anno_data[order(factor(anno_data[,1], levels = levels(Both_Genome$V1)), anno_data[,2], anno_data[,3]),]

B_hg <- h_data[,1:3]
B_mm <- allHMLG_data[,c(7:9)]

## This is to format the data generate here as a supplementary file for the paper:
toSupp <- allHMLG_data
colnames(toSupp) <- c("mgi_symbol", "chr.hg", "start.hg", "end.hg", "hgnc_symbol", "percent_homology", "chr.mm", "qtl.lowCI", "qtl.highCI", "lodcolumn", "qtl", "lod","distanceToTSS")

toSupp <- merge(toSupp, mod_data, by=c("chr.mm", "qtl.lowCI", "qtl.highCI", "qtl"))
toSupp <- toSupp[,c(1,4,2:3,5,9,6:8,12,14,16)]
toSupp <- toSupp[order(toSupp[,2]),]
toSupp[,2] <- round(as.numeric(toSupp[,2]) / 1000000)
toSupp[,8] <- round(as.numeric(toSupp[,8]) / 1000000)
toSupp <- toSupp[,c(5, 6, 7, 8, 1,3,4,2,10,11,12)]
colnames(toSupp) <- c("nearest_gene_mgi","syntenic_gene_hg","syntenic_hg19_chr", "syntenic_hg19_Mbp", "mm10_chr", "qtl.lowCI", "qtl.highCI", "qtl_Mbp", "lod.score", "trait", "enrichment")
toSupp[,3] <- paste0("hg19_", toSupp[,3])
toSupp[,9] <- round(as.numeric(toSupp[,9]), 2)
# write.table(toSupp, file="tableInfo_moduleTraits_forSupplement.csv", quote=F, row.names=F, sep=",")
modData <- toSupp
```

Plotting the module data circle plot

```{r, cache=TRUE, fig.height=20, fig.width=20, message=FALSE}
set.seed(2020)
colfunc <- colorRampPalette(c("paleturquoise1", "blue"))
COL1 <- "#a8df49"
colfunc2 <- colorRampPalette(c("orange", "red"))
COL2 <- colfunc2(21)[2]

circos.clear()

circos.par(start.degree = 85, "gap.degree" = c(rep(1, 22), 10, rep(1, 19), 10),
           canvas.xlim=c(-.1,.1), canvas.ylim=c(-.95, .95), track.margin=c(.01, .01),
           cell.padding=c(.01, .01, .01, .01), track.height = 0.3, points.overflow.warning = FALSE)

circos.initializeWithIdeogram(plotType=NULL, Both_Genome, sector.width = sector.width, 
                              sort.chr = FALSE, track.height=.8)
circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = .2)
circos.genomicTrack(plotType1, track.index=1, bg.border="white", ylim=c(0,20),
                    panel.fun = function(region, value, ...) {
                      circos.genomicLines(region, value, col = "brown3", type='l', lwd=4)
                    })
circos.yaxis(side = "left", sector.index = "chr1", track.index=1, labels.cex=1.1, col="black", labels.col="black")
circos.genomicTrack(plotType2, track.index=1, bg.border="white", ylim=c(0,20),
                    panel.fun = function(region, value, ...) {
                      circos.genomicLines(region, value, col = "dodgerblue", type='l', lwd=3)
                    })
circos.genomicTrack(hotspotGenes, track.index=1, stack=T, bg.border="white",
                    numeric.column = 4, 
                    panel.fun = function(region, value, ...) { 
                      circos.genomicLines(region, value, col = "red", type="h", lty=2)
                    })
circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = .8)

anno_data$cols1[anno_data$cols1=="white"] <- "black"
circos.genomicLabels(anno_data, labels.column = 4, cex=1.4, track.margin=c(0,0), 
                     side="outside", padding=.1, col="black", line_col= anno_data[,5],
                     connection_height=convert_height(10, "mm"), line_lwd=4,
                     labels_height = (convert_height(1, "cm")))

circos.par(track.margin=c(.01, .01),
           cell.padding=c(.001, .001, .001, .001), track.height = 0.1)

COLORS <- c(rep(COL1, 23), rep(COL2, 20))
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
  chr = CELL_META$sector.index
  xlim = CELL_META$xlim
  ylim = CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col = COLORS[CELL_META$sector.numeric.index], border="white")
  circos.text(mean(xlim), mean(ylim), NAMES[CELL_META$sector.numeric.index], col = "black", cex=1.5,
              facing="downward", niceFacing = TRUE)
}, track.height = 0.25, bg.border = NA)

circos.par("track.height" = 0.1, cell.padding = c(.1, .1, .1, .1))
X <- anno_data[78:nrow(anno_data),]
X <- X[order(X$chr),]
circos.genomicLink(B_mm, B_hg, col=X[,5])


LABELS <- c()
for(i in 1:nrow(mod_data)) {
  LABELS[i] <- paste0(mod_data[i,7], " (", mod_data[i,1], ")")
}

lgd_points = Legend(at = unique(LABELS), type = "lines", nrow=6, ncol=6, border=NA, 
                    background="white", grid_height=unit(4, "mm"), grid_width= unit(4, "mm"),
                    legend_gp = gpar(col = unique(mod_data[,1]), lwd=5, lineend="square"), 
                    labels_gp = gpar(fontsize = 9), title_position = "topleft",
                    title = "")

pushViewport(viewport(x = unit(2, "mm"), y = unit(2, "mm"),
                      width = grobWidth(lgd_points),
                      height = grobHeight(lgd_points),
                      just = c("left", "bottom")))
grid.draw(lgd_points)
upViewport()

```



## Enrichment Analysis:

```{r}

# Use the RNA-seq data to exclude genes from consideration which are the lowest expressed.
rnaseqdata <- read.table("texpr_mrna.csv", header=T, 
              stringsAsFactors=F, sep=",", row.names=1)
gmeans <- rowMeans(rnaseqdata[,-c(1:5)])
userna <- rnaseqdata[which(gmeans >= quantile(gmeans, .05)),]
exprg <- rownames(userna)

# Only include those genes and also those on chromsomes 1 - 22, and X in Human (as in our analysis above).
orth.human <- subset(orth.2, ensembl_gene_id %in% exprg)
orth.human <- subset(orth.human, chromosome_name %in% c(as.character(1:22), "X") & hgnc_symbol!="")

# Unique human genes
orth.human <- orth.human[,4:8]
orth.human <- orth.human[!duplicated(orth.human),-1]

# Get human locations orthologous to our QTL.
ALLTraits <- clinData[,7:8]
ALLTraits <- rbind(ALLTraits, modData[,3:4])
ALLTraits[,2] <- as.numeric(ALLTraits[,2])*1e6

#Only use unique ones:
ALLTraits <- ALLTraits[!duplicated(ALLTraits),]

# For each the the 97 qtls, figure out which chromosome it is on, then
# using the FULL list of diab SNPS that was downloaded,
# check if it is within 1e6 bases.
diab.both <- rbind(diab.type1, diab.type2)
diab.both <- diab.both[!duplicated(diab.both[,1:4]),]

numIn.both <- 0
for(i in 1:nrow(ALLTraits)) {
  #first needs to be on the same chrom.
  CHR <- gsub("hg19_", "", ALLTraits[i,1])
  useSS <- subset(diab.both, chr == CHR)
#now check if near any T2 or T1 snps.
  if(any(abs(as.numeric(useSS[,2]) - as.numeric(ALLTraits[i,2])) < 1e6)) {
    numIn.both <- numIn.both + 1
  }
}
numIn.both

# Now counts how many 'random' human locations are near the set of diabetes SNPs
set.seed(1)
numIn <- c()
for(j in 1:1000) {
  ss.hg <- orth.human[sample(1:nrow(orth.human), nrow(ALLTraits)),]

  numIn[j] <- 0
  for(i in 1:nrow(ss.hg)) {
  
  #first needs to be on the same chrom.
    CHR <- paste0("chr", ss.hg[i,1])
    useSS <- subset(diab.both, chr == CHR)
  
    if(any(abs(as.numeric(useSS[,2]) - ss.hg[i,3]) < 1e6)) {
      numIn[j] <- numIn[j] + 1
    }
  }
}
# P-value
pval <- (sum(numIn >= (numIn.both)) +1 )/ (length(numIn)+1); pval
# Plot histogram
hist(numIn /90, xlab="Proportion of human genes within 1Mb of a diabetes GWAS SNP \n (out of 90)", 
	breaks=10, xlim=c(0,.6), ylim=c(0, 500), main="", 
  cex.axis=1.2, cex.lab=1.2)
abline(v = (numIn.both/90), lwd=3)


```