---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 03-load-explore-data.md in _episodes_rmd/
title: "Load and explore the data"
teaching: 15
exercises: 30
questions:
- "What data are required for eqtl mapping?"
objectives:
- To provide an example and exploration of data used for eqtl mapping.
keypoints:
- ""
source: Rmd
---



Load the libraries.

~~~
library(tidyverse)
library(knitr)
library(GGally)
library(corrplot)
library(broom)
library(qtl2)
library(qtl2convert)
library(qtl2ggplot)
library(RColorBrewer)
# the following analysis is from File S1 Attie_eQTL_paper_physiology.Rmd 
# compliments of Daniel Gatti. See Data Dryad entry for more information.
~~~
{: .language-r}

## Physiological Phenotypes

The complete data used in these analyses are available from 
[Data Dryad](https://doi.org/10.5061/dryad.pj105). 

Load in the clinical phenotypes.


~~~
# load the data
load("../data/attie_DO500_clinical.phenotypes.RData")
~~~
{: .language-r}

See the [data dictionary](../data/Attie-232_Attie_DO_Islets-dictionary.csv) to 
see a description of each of these phenotypes. You can also view a table of
the data dictionary.


~~~
kable(pheno_clin_dict)
~~~
{: .language-r}



|                   |name                     |short_name         |pheno_type  |is_numeric |is_derived |formula                          |description                                                                                                                                                                                                                                             |
|:------------------|:------------------------|:------------------|:-----------|:----------|:----------|:--------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|mouse              |Mouse                    |mouse              |demographic |FALSE      |FALSE      |NA                               |Animal identifier.                                                                                                                                                                                                                                      |
|sex                |sex                      |sex                |demographic |FALSE      |FALSE      |NA                               |Male (M) or female (F).                                                                                                                                                                                                                                 |
|sac_date           |date of sac              |sac_date           |demographic |FALSE      |FALSE      |NA                               |Date when mouse was sacrificed; used to compute days on diet, using birth dates.                                                                                                                                                                        |
|partial_inflation  |partial inflation        |partial_inflation  |demographic |FALSE      |FALSE      |NA                               |Some mice showed a partial pancreatic inflation which would negatively effect the total number of islets collected from these mice.                                                                                                                     |
|coat_color         |coat color               |coat_color         |demographic |FALSE      |FALSE      |NA                               |Visual inspection by Kathy Schuler on coat color.                                                                                                                                                                                                       |
|oGTT_date          |oGTT date                |oGTT_date          |demographic |FALSE      |FALSE      |NA                               |Date the oGTT was performed.                                                                                                                                                                                                                            |
|FAD_NAD_paired     |FAD_NAD Paired/Unpaired  |FAD_NAD_paired     |demographic |FALSE      |FALSE      |NA                               |A change in the method that was used to make this measurement by Matt Merrins' lab. Paired was the same islet for the value at 3.3mM vs. 8.3mM glucose; unpaired was where averages were used for each glucose concentration and used to compute ratio. |
|FAD_NAD_filter_set |FAD_NAD FAD filter set   |FAD_NAD_filter_set |demographic |FALSE      |FALSE      |NA                               |A different filter set was used on the microscope to make the fluorescent measurement; may have influenced the values.                                                                                                                                  |
|crumblers          |Crumblers                |crumblers          |demographic |FALSE      |FALSE      |NA                               |Some mice store food in their bedding (hoarders) which would be incorrectly interpreted as consumed.                                                                                                                                                    |
|birthdate          |Birth date               |birthdate          |demographic |FALSE      |FALSE      |NA                               |Birth date                                                                                                                                                                                                                                              |
|diet_days          |days on diet             |diet_days          |clinical    |TRUE       |FALSE      |NA                               |Number of days.                                                                                                                                                                                                                                         |
|num_islets         |number of islets         |num_islets         |clinical    |TRUE       |FALSE      |NA                               |Total number of islets harvested per mouse; negatively impacted by those with partial inflation.                                                                                                                                                        |
|Ins_per_islet      |Ins content ng per islet |Ins_per_islet      |clinical    |TRUE       |FALSE      |NA                               |Amount of insulin per islet in units of ng/ml/islet.                                                                                                                                                                                                    |
|WPIC               |WPIC estimate            |WPIC               |clinical    |TRUE       |TRUE       |Ins_per_islet * num_islets       |Derived number; equal to total number of islets times insulin content per islet.                                                                                                                                                                        |
|HOMA_IR_0min       |HOMA-IR at 0min          |HOMA_IR_0min       |clinical    |TRUE       |TRUE       |Glu_0min * Ins_0min / 405        |glucose*insulin/405 at time t=0 for the oGTT                                                                                                                                                                                                            |
|HOMA_B_0min        |HOMA-B at 0min           |HOMA_B_0min        |clinical    |TRUE       |TRUE       |360 * Ins_0min / (Glu_0min - 63) |360 * Insulin / (Glucose - 63) at time t=0 for the oGTT                                                                                                                                                                                                 |
|Glu_tAUC           |glucose tAUC             |Glu_tAUC           |clinical    |TRUE       |TRUE       |complicated                      |Area under the curve (AUC) calculation without any correction for baseline differences.                                                                                                                                                                 |
|Ins_tAUC           |insulin tAUC             |Ins_tAUC           |clinical    |TRUE       |TRUE       |complicated                      |Area under the curve (AUC) calculation without any correction for baseline differences.                                                                                                                                                                 |
|Glu_6wk            |6 wk glu                 |Glu_6wk            |clinical    |TRUE       |FALSE      |NA                               |Plasma glucose with units of mg/dl; fasting.                                                                                                                                                                                                            |
|Ins_6wk            |6 wk ins                 |Ins_6wk            |clinical    |TRUE       |FALSE      |NA                               |Plasma insulin with units of ng/ml; fasting.                                                                                                                                                                                                            |
|TG_6wk             |6 wk TG                  |TG_6wk             |clinical    |TRUE       |FALSE      |NA                               |Plasma triglyceride (TG) with units of mg/dl; fasting.                                                                                                                                                                                                  |
|Glu_10wk           |10 wk glu                |Glu_10wk           |clinical    |TRUE       |FALSE      |NA                               |Plasma glucose with units of mg/dl; fasting.                                                                                                                                                                                                            |
|Ins_10wk           |10 wk ins                |Ins_10wk           |clinical    |TRUE       |FALSE      |NA                               |Plasma insulin with units of ng/ml; fasting.                                                                                                                                                                                                            |
|TG_10wk            |10 wk TG                 |TG_10wk            |clinical    |TRUE       |FALSE      |NA                               |Plasma triglyceride (TG) with units of mg/dl; fasting.                                                                                                                                                                                                  |
|Glu_14wk           |14 wk glu                |Glu_14wk           |clinical    |TRUE       |FALSE      |NA                               |Plasma glucose with units of mg/dl; fasting.                                                                                                                                                                                                            |
|Ins_14wk           |14 wk ins                |Ins_14wk           |clinical    |TRUE       |FALSE      |NA                               |Plasma insulin with units of ng/ml; fasting.                                                                                                                                                                                                            |
|TG_14wk            |14 wk TG                 |TG_14wk            |clinical    |TRUE       |FALSE      |NA                               |Plasma triglyceride (TG) with units of mg/dl; fasting.                                                                                                                                                                                                  |
|food_ave           |Avg gm per day           |food_ave           |clinical    |TRUE       |TRUE       |complicated                      |Average food consumption over the measurements made for each mouse.                                                                                                                                                                                     |
|weight_2wk         |body weight week 2       |weight_2wk         |body weight |TRUE       |FALSE      |NA                               |Body weight at indicated date; units are gm.                                                                                                                                                                                                            |
|weight_6wk         |body weight week 6       |weight_6wk         |body weight |TRUE       |FALSE      |NA                               |Body weight at indicated date; units are gm.                                                                                                                                                                                                            |
|weight_10wk        |body weight week 10      |weight_10wk        |body weight |TRUE       |FALSE      |NA                               |Body weight at indicated date; units are gm.                                                                                                                                                                                                            |
|DOwave             |DO wave                  |DOwave             |demographic |TRUE       |FALSE      |NA                               |Wave (i.e., batch) of DO mice                                                                                                                                                                                                                           |


~~~
names(pheno_clin)
~~~
{: .language-r}



~~~
 [1] "mouse"              "sex"                "sac_date"          
 [4] "partial_inflation"  "coat_color"         "oGTT_date"         
 [7] "FAD_NAD_paired"     "FAD_NAD_filter_set" "crumblers"         
[10] "birthdate"          "diet_days"          "num_islets"        
[13] "Ins_per_islet"      "WPIC"               "HOMA_IR_0min"      
[16] "HOMA_B_0min"        "Glu_tAUC"           "Ins_tAUC"          
[19] "Glu_6wk"            "Ins_6wk"            "TG_6wk"            
[22] "Glu_10wk"           "Ins_10wk"           "TG_10wk"           
[25] "Glu_14wk"           "Ins_14wk"           "TG_14wk"           
[28] "food_ave"           "weight_2wk"         "weight_6wk"        
[31] "weight_10wk"        "DOwave"            
~~~
{: .output}


~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
~~~
{: .language-r}

### Figure 1 Boxplots


~~~
pheno_clin %>%
  select(mouse, sex, starts_with("weight")) %>%
  gather(week, value, -mouse, -sex) %>%
  separate(week, c("tmp", "week")) %>%
  mutate(week = factor(week, levels = c("2wk", "6wk", "10wk"))) %>%
  ggplot(aes(week, value, fill = sex)) +
    geom_boxplot() +
    scale_y_log10() +
    labs(title = "Body Weight", y = "Weight")
~~~
{: .language-r}

<img src="../fig/rmd-03-bw_boxplot-1.png" alt="plot of chunk bw_boxplot" width="612" style="display: block; margin: auto;" />


~~~
pheno_clin %>%
  select(mouse, sex, starts_with("Glu")) %>%
  select(mouse, sex, ends_with("wk")) %>%
  gather(week, value, -mouse, -sex) %>%
  separate(week, c("tmp", "week")) %>%
  mutate(week = factor(week, levels = c("6wk", "10wk", "14wk"))) %>%
  ggplot(aes(week, value, fill = sex)) +
    geom_boxplot() +
    scale_y_log10() +
    labs(title = "Glucose", y = "Glucose")
~~~
{: .language-r}

<img src="../fig/rmd-03-glucose_boxplot-1.png" alt="plot of chunk glucose_boxplot" width="612" style="display: block; margin: auto;" />


~~~
pheno_clin %>%
  select(mouse, sex, starts_with("Ins")) %>%
  select(mouse, sex, ends_with("wk")) %>%
  gather(week, value, -mouse, -sex) %>%
  separate(week, c("tmp", "week")) %>%
  mutate(week = factor(week, levels = c("6wk", "10wk", "14wk"))) %>%
  ggplot(aes(week, value, fill = sex)) +
    geom_boxplot() +
    scale_y_log10() +
    labs(title = "Insulin", y = "Insulin")
~~~
{: .language-r}

<img src="../fig/rmd-03-insulin_boxplot-1.png" alt="plot of chunk insulin_boxplot" width="612" style="display: block; margin: auto;" />


~~~
pheno_clin %>%
  select(mouse, sex, starts_with("TG")) %>%
  select(mouse, sex, ends_with("wk")) %>%
  gather(week, value, -mouse, -sex) %>%
  separate(week, c("tmp", "week")) %>%
  mutate(week = factor(week, levels = c("6wk", "10wk", "14wk"))) %>%
  ggplot(aes(week, value, fill = sex)) +
    geom_boxplot() +
    scale_y_log10() +
    labs(title = "TG", y = "TG")
~~~
{: .language-r}

<img src="../fig/rmd-03-trig_boxplot-1.png" alt="plot of chunk trig_boxplot" width="612" style="display: block; margin: auto;" />


~~~
pheno_clin %>%
  select(mouse, sex, num_islets:Ins_tAUC, food_ave) %>%
  gather(phenotype, value, -mouse, -sex) %>%
  ggplot(aes(sex, value, fill = sex)) +
    geom_boxplot() +
    scale_y_log10() +
    facet_wrap(~phenotype, scales = "free_y")
~~~
{: .language-r}

<img src="../fig/rmd-03-fig1_boxplots-1.png" alt="plot of chunk fig1_boxplots" width="612" style="display: block; margin: auto;" />

### QA/QC

Log transform and standardize each phenotype. Consider setting points that are 
more than 5 std. dev. from the mean to NA. Only do this if the final 
distribution doesn't look skewed.


~~~
pheno_clin_log = pheno_clin %>%
                   mutate_if(is.numeric, log)
pheno_clin_std = pheno_clin_log %>%
                   select(mouse, num_islets:weight_10wk) %>%
                   mutate_if(is.numeric, scale)
pheno_clin_std %>%
  select(num_islets:weight_10wk) %>%
  gather(phenotype, value) %>%
  ggplot(aes(x = phenotype, y = value)) +
    geom_boxplot() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
    labs(title = "Distribution of Standardized Phenotypes")
~~~
{: .language-r}

<img src="../fig/rmd-03-pheno_std-1.png" alt="plot of chunk pheno_std" width="612" style="display: block; margin: auto;" />

### Tests for sex, wave and diet_days.


~~~
tmp = pheno_clin_log %>%
        select(mouse, sex, DOwave, diet_days, num_islets:weight_10wk) %>%
        gather(phenotype, value, -mouse, -sex, -DOwave, -diet_days) %>%
        group_by(phenotype) %>%
        nest()
mod_fxn = function(df) {
  lm(value ~ sex + DOwave + diet_days, data = df)
}
tmp = tmp %>%
  mutate(model = map(data, mod_fxn)) %>%
  mutate(summ = map(model, tidy)) %>%
  unnest(summ)
# kable(tmp, caption = "Effects of Sex, Wave & Diet Days on Phenotypes")
~~~
{: .language-r}


~~~
tmp %>%
  filter(term != "(Intercept)") %>%
  mutate(neg.log.p = -log10(p.value)) %>%
  ggplot(aes(term, neg.log.p)) +
    geom_point() +
    facet_wrap(~phenotype) +
    labs(title = "Significance of Sex, Wave & Diet Days on Phenotypes") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
rm(tmp)
~~~
{: .language-r}

<img src="../fig/rmd-03-sex_diet_wave_effects-1.png" alt="plot of chunk sex_diet_wave_effects" width="612" style="display: block; margin: auto;" />

### Weight vs. Food Intake


~~~
pheno_clin_log %>%
  select(mouse, sex, food_ave:weight_10wk) %>%
  gather(phenotype, value, -mouse, -sex, -food_ave) %>%
  separate(phenotype, c("phenotype", "week")) %>%
  mutate(week = factor(week, levels = c("2wk", "6wk", "10wk"))) %>%
  ggplot(aes(food_ave, value, color = sex)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(title = "Food Intake vs. Body Weight", y = "log(Body Weight)") + 
    facet_wrap(~week)
~~~
{: .language-r}



~~~
`geom_smooth()` using formula 'y ~ x'
~~~
{: .output}

<img src="../fig/rmd-03-bw_vs_food-1.png" alt="plot of chunk bw_vs_food" width="612" style="display: block; margin: auto;" />


~~~
model_fxn = function(df) { lm(value ~ sex*food_ave, data = df) }
tmp = pheno_clin_log %>%
  select(mouse, sex, food_ave:weight_10wk) %>%
  gather(phenotype, value, -mouse, -sex, -food_ave) %>%
  separate(phenotype, c("phenotype", "week")) %>%
  mutate(week = factor(week, levels = c("2wk", "6wk", "10wk"))) %>%
  group_by(week) %>%
  nest() %>%
  mutate(model = map(data, model_fxn)) %>%
  mutate(summ = map(model, tidy)) %>%
  unnest(summ) %>%
  filter(term != "(Intercept)") %>%
  mutate(p.adj = p.adjust(p.value))
# kable(tmp, caption = "Effects of Sex and Food Intake on Body Weight")
~~~
{: .language-r}

### Correlation Plots

Females


~~~
tmp = pheno_clin_log %>% 
        filter(sex == "F") %>%
        select(num_islets:weight_10wk)
tmp = cor(tmp, use = "pairwise")
corrplot.mixed(tmp, upper = "ellipse", lower = "number", 
               main = "Female Clinical Phenotype Correlation")
corrplot.mixed(tmp, upper = "ellipse", lower = "number", 
               main = "Female Clinical Phenotype Correlation")
~~~
{: .language-r}

<img src="../fig/rmd-03-female_corr_plot-1.png" alt="plot of chunk female_corr_plot" width="1080" style="display: block; margin: auto;" />

Males


~~~
tmp = pheno_clin_log %>% 
        filter(sex == "M") %>%
        select(num_islets:weight_10wk)
tmp = cor(tmp, use = "pairwise")
corrplot.mixed(tmp, upper = "ellipse", lower = "number", 
               main = "Male Clinical Phenotype Correlation")
corrplot.mixed(tmp, upper = "ellipse", lower = "number", 
               main = "Male Clinical Phenotype Correlation")
~~~
{: .language-r}

<img src="../fig/rmd-03-male_corr_plot-1.png" alt="plot of chunk male_corr_plot" width="1080" style="display: block; margin: auto;" />


## Gene Expression Phenotypes

### summarized data, matrices w/sample annotation, exp data, gene annotation, 
### like bioconductor summarized experiment etc = whole set of rectangles


~~~
# load the expression data along with annotations and metadata
load("../data/dataset.islet.rnaseq.RData")
~~~
{: .language-r}



~~~
Warning in readChar(con, 5L, useBytes = TRUE): cannot open compressed file '../
data/dataset.islet.rnaseq.RData', probable reason 'No such file or directory'
~~~
{: .warning}



~~~
Error in readChar(con, 5L, useBytes = TRUE): cannot open the connection
~~~
{: .error}



~~~
# look at raw counts
dataset.islet.rnaseq$raw[1:6,1:6]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at gene annotations
dataset.islet.rnaseq$annots[1:6,]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at sample metadata
# mouse sex, birth dates and DO waves
table(dataset.islet.rnaseq$samples[, c("sex", "birthdate")])
~~~
{: .language-r}



~~~
Error in table(dataset.islet.rnaseq$samples[, c("sex", "birthdate")]): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
table(dataset.islet.rnaseq$samples[, c("sex", "DOwave")])
~~~
{: .language-r}



~~~
Error in table(dataset.islet.rnaseq$samples[, c("sex", "DOwave")]): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at raw counts
dataset.islet.rnaseq$raw[1:6,1:6]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at normalized counts
dataset.islet.rnaseq$expr[1:6,1:6]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at LOD peaks
dataset.islet.rnaseq$lod.peaks[1:6,]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at a region of interest
chr11_peaks <- dataset.islet.rnaseq$annots %>% 
   select(gene_id, chr) %>% 
   filter(chr=="11") %>%
   left_join(dataset.islet.rnaseq$lod.peaks, 
             by = c("chr" = "chrom", "gene_id" = "annot.id")) 
~~~
{: .language-r}



~~~
Error in select(., gene_id, chr): object 'dataset.islet.rnaseq' not found
~~~
{: .error}



~~~
# look at the first several rows of chromosome 11 peaks
head(chr11_peaks)
~~~
{: .language-r}



~~~
Error in head(chr11_peaks): object 'chr11_peaks' not found
~~~
{: .error}



~~~
# how many rows?
dim(chr11_peaks)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'chr11_peaks' not found
~~~
{: .error}



~~~
# how many rows have LOD scores?
chr11_peaks %>% filter(!is.na(lod)) %>% dim()
~~~
{: .language-r}



~~~
Error in filter(., !is.na(lod)): object 'chr11_peaks' not found
~~~
{: .error}



~~~
# sort chromosome 11 peaks by LOD score
chr11_peaks %>% arrange(desc(lod)) %>% head()
~~~
{: .language-r}



~~~
Error in arrange(., desc(lod)): object 'chr11_peaks' not found
~~~
{: .error}



~~~
# range of LOD scores and positions
range(chr11_peaks$lod, na.rm = TRUE)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'chr11_peaks' not found
~~~
{: .error}



~~~
range(chr11_peaks$pos, na.rm = TRUE)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'chr11_peaks' not found
~~~
{: .error}
