---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 06-map-many-eqtls.md in _episodes_rmd/
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "To map several genes at the same time"
source: Rmd
---



### Load Libraries  


~~~
library(tidyverse)
library(knitr)
library(broom)
library(qtl2)
library(qtl2ggplot)
library(RColorBrewer)

source("../code/gg_transcriptome_map.R")
source("../code/qtl_heatmap.R")
~~~
{: .language-r}

Before we begin this lesson, we need to create another directory called `results` in our main directory.  You can do this by clicking on the "Files" tab and navigate into the main directory.  Then select "New Folder" and name it "results".

### Load Data


~~~
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# data from paper
load("../data/dataset.islet.rnaseq.RData")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

# genotype probabilities
probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")
~~~
{: .language-r}

### Data Selection

For this lesson, lets choose a random set of 50 gene expression phenotypes.




~~~
genes = colnames(norm)

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[sams]

gene.info <- dataset.islet.rnaseq$annots[genes,]
rownames(gene.info) = NULL
kable(gene.info)
~~~
{: .language-r}



|gene_id            |symbol        |chr |      start|        end| strand|    middle|nearest.marker.id |biotype              |module          |hotspot |
|:------------------|:-------------|:---|----------:|----------:|------:|---------:|:-----------------|:--------------------|:---------------|:-------|
|ENSMUSG00000082855 |Gm14537       |X   |  20.391709|  20.392501|      1|  20.39211|X_20409194        |pseudogene           |grey            |NA      |
|ENSMUSG00000053889 |A130066N16Rik |9   |  34.918233|  34.929509|     -1|  34.92387|9_34976867        |antisense            |grey            |NA      |
|ENSMUSG00000087002 |Gm16277       |19  |  58.204139|  58.236047|      1|  58.22009|19_58222556       |antisense            |grey            |NA      |
|ENSMUSG00000072692 |Gm13826       |5   | 115.102923| 115.110268|     -1| 115.10660|5_115033401       |pseudogene           |sienna3         |NA      |
|ENSMUSG00000026778 |Prkcq         |2   |  11.172113|  11.301226|      1|  11.23667|2_11207176        |protein_coding       |pink            |NA      |
|ENSMUSG00000041120 |Nbl1          |4   | 139.082284| 139.092993|     -1| 139.08764|4_139107368       |protein_coding       |yellow          |NA      |
|ENSMUSG00000093059 |Gm24809       |X   |  37.071992|  37.072262|     -1|  37.07213|X_37022542        |misc_RNA             |grey            |NA      |
|ENSMUSG00000058812 |0610039K10Rik |2   | 163.644850| 163.645801|      1| 163.64533|2_163582672       |antisense            |grey            |NA      |
|ENSMUSG00000028411 |Aptx          |4   |  40.682382|  40.703194|     -1|  40.69279|4_40670078        |protein_coding       |green           |NA      |
|ENSMUSG00000029591 |Ung           |5   | 114.130386| 114.139323|      1| 114.13485|5_114111853       |protein_coding       |grey            |NA      |
|ENSMUSG00000045733 |Sprn          |7   | 140.150628| 140.154877|     -1| 140.15275|7_140163958       |protein_coding       |paleturquoise   |NA      |
|ENSMUSG00000050982 |Apol10a       |15  |  77.477047|  77.491069|      1|  77.48406|15_77426124       |protein_coding       |grey            |NA      |
|ENSMUSG00000042647 |Acad12        |5   | 121.598284| 121.618938|     -1| 121.60861|5_121660329       |protein_coding       |grey            |NA      |
|ENSMUSG00000061778 |Mospd2        |X   | 164.936169| 164.980375|     -1| 164.95827|X_164962098       |protein_coding       |brown           |NA      |
|ENSMUSG00000004096 |Cwc15         |9   |  14.500617|  14.510577|      1|  14.50560|9_14501195        |protein_coding       |brown           |NA      |
|ENSMUSG00000045374 |Wdr81         |11  |  75.440944|  75.454717|     -1|  75.44783|11_75466326       |protein_coding       |yellow          |NA      |
|ENSMUSG00000058444 |Map2k5        |9   |  63.163769|  63.377902|     -1|  63.27084|9_63267626        |protein_coding       |yellow          |NA      |
|ENSMUSG00000038152 |5033430I15Rik |13  |  45.965253|  45.966500|      1|  45.96588|13_45964061       |protein_coding       |red             |NA      |
|ENSMUSG00000063018 |2010204K13Rik |X   |   7.411817|   7.445143|     -1|   7.42848|X_7440367         |processed_transcript |blue            |NA      |
|ENSMUSG00000073016 |Uprt          |X   | 104.482818| 104.507490|      1| 104.49515|X_104495767       |protein_coding       |darkturquoise   |NA      |
|ENSMUSG00000028221 |Tmem55a       |4   |  14.864076|  14.915176|      1|  14.88963|4_14880547        |protein_coding       |turquoise       |NA      |
|ENSMUSG00000029368 |Alb           |5   |  90.460889|  90.476603|      1|  90.46875|5_90479033        |protein_coding       |grey            |NA      |
|ENSMUSG00000005198 |Polr2a        |11  |  69.733997|  69.758637|     -1|  69.74632|11_69733033       |protein_coding       |darkturquoise   |NA      |
|ENSMUSG00000054659 |Pm20d2        |4   |  33.174230|  33.189737|     -1|  33.18198|4_33245486        |protein_coding       |yellow          |NA      |
|ENSMUSG00000020260 |Pofut2        |10  |  77.259300|  77.269586|      1|  77.26444|10_77384449       |protein_coding       |black           |NA      |
|ENSMUSG00000063430 |Wscd2         |5   | 113.490333| 113.589725|      1| 113.54003|5_113653421       |protein_coding       |turquoise       |NA      |
|ENSMUSG00000020580 |Rock2         |12  |  16.894978|  16.988274|      1|  16.94163|12_16893631       |protein_coding       |pink            |NA      |
|ENSMUSG00000030386 |Zfp606        |7   |  12.478293|  12.496235|      1|  12.48726|7_12326280        |protein_coding       |green           |NA      |
|ENSMUSG00000033177 |Tmprss7       |16  |  45.656315|  45.693658|     -1|  45.67499|16_45673536       |protein_coding       |grey            |NA      |
|ENSMUSG00000066513 |Klk1b4        |7   |  44.207435|  44.211754|      1|  44.20959|7_44138903        |protein_coding       |darkgrey        |NA      |
|ENSMUSG00000031256 |Cstf2         |X   | 134.059187| 134.086819|      1| 134.07300|X_134071661       |protein_coding       |turquoise       |NA      |
|ENSMUSG00000048612 |Myof          |19  |  37.899036|  38.043577|     -1|  37.97131|19_37953577       |protein_coding       |grey            |NA      |
|ENSMUSG00000093751 |Gm3307        |7   |  63.606019|  63.606650|      1|  63.60633|7_63722318        |pseudogene           |grey            |NA      |
|ENSMUSG00000059742 |Kcnh7         |2   |  62.693414|  63.184287|     -1|  62.93885|2_62884344        |protein_coding       |turquoise       |NA      |
|ENSMUSG00000037845 |Fdxacb1       |9   |  50.768236|  50.773348|      1|  50.77079|9_50740446        |protein_coding       |turquoise       |NA      |
|ENSMUSG00000020321 |Mdh1          |11  |  21.556787|  21.572367|     -1|  21.56458|11_21569820       |protein_coding       |salmon          |NA      |
|ENSMUSG00000031384 |Asb9          |X   | 164.506327| 164.539752|      1| 164.52304|X_164552669       |protein_coding       |brown           |NA      |
|ENSMUSG00000086487 |Gm11638       |11  | 105.363252| 105.390642|      1| 105.37695|11_105410419      |processed_transcript |grey            |NA      |
|ENSMUSG00000034683 |Ppp1r1c       |2   |  79.707780|  79.818496|      1|  79.76314|2_79904938        |protein_coding       |grey            |NA      |
|ENSMUSG00000001349 |Cnn1          |9   |  22.099281|  22.109630|      1|  22.10446|9_22111151        |protein_coding       |red             |NA      |
|ENSMUSG00000022370 |Mrpl13        |15  |  55.534094|  55.557748|     -1|  55.54592|15_55579677       |protein_coding       |salmon          |NA      |
|ENSMUSG00000034269 |Setd5         |6   | 113.077365| 113.153435|      1| 113.11540|6_112982487       |protein_coding       |saddlebrown     |NA      |
|ENSMUSG00000052616 |Ccdc79        |8   | 104.446719| 104.509910|     -1| 104.47831|8_104460981       |protein_coding       |grey            |NA      |
|ENSMUSG00000078868 |Gm14412       |2   | 177.314520| 177.324307|     -1| 177.31941|2_177282713       |protein_coding       |lightsteelblue1 |NA      |
|ENSMUSG00000033597 |Caskin1       |17  |  24.488783|  24.508905|      1|  24.49884|17_24471676       |protein_coding       |yellow          |NA      |
|ENSMUSG00000093514 |Gm9320        |17  |  70.853767|  70.876227|      1|  70.86500|17_70935012       |pseudogene           |grey            |NA      |
|ENSMUSG00000035311 |Gnptab        |10  |  88.379132|  88.447329|      1|  88.41323|10_88509852       |protein_coding       |steelblue       |NA      |
|ENSMUSG00000022336 |Eif3e         |15  |  43.250040|  43.282736|     -1|  43.26639|15_43369640       |protein_coding       |salmon          |NA      |
|ENSMUSG00000097707 |Gm26742       |18  |  54.981435|  54.987689|      1|  54.98456|18_54969054       |lincRNA              |grey            |NA      |
|ENSMUSG00000021668 |Polk          |13  |  96.480689|  96.542485|     -1|  96.51159|13_96509439       |protein_coding       |green           |NA      |

### Expression Data

Lets check the distribution for the first 20 gene expression phenotypes. If you would like to check the distribution of all 50 genes, change `for(gene in genes[1:20])` in the code below to `for(gene in genes)`.


~~~
par(mfrow=c(3,4))
for(gene in genes[1:20]){
  hist(norm[,gene], main = gene)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-hist_untransformed-2.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

Check the distributions.  Do they all have a normal distribution?

You will notice that the distribtion of some genes are skewed to the left.  This means that that only a small amount of samples have data and therefore, will need to be removed. 


~~~
limit = nrow(norm) * 0.25

genes0 = which(as.numeric(colSums(counts[,genes] == 0)) > limit) 

genes <- genes[-genes0]
~~~
{: .language-r}

### The Marker Map  

We are using the same marker map as in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#the-marker-map)


### Genotype probabilities  

We have explored this earlier in th previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#genotype-probabilities).  But, as a reminder, we have already calculated genotype probabilities which we loaded above called `probs`.  This contains the 8 state genotype probabilities using the 69k grid  map of the same 500 DO mice that also have clinical phenotypes. 


### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

We have explored the kinship matrix in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#kinship-matrix). It has already been calculated and loaded in above. 


### Covariates    

Now let's add the necessary covariates. For these 50 gene expression data, we will correct for `DOwave`,`sex` and `diet_days`.


~~~
###merging covariate data and expression data to test for sex, wave and diet_days.

cov.counts <- merge(covar, norm[,genes], by=c("row.names"), sort=F)

#testing covairates on expression data

tmp = cov.counts %>%
        dplyr::select(mouse, sex, DOwave, diet_days, names(cov.counts[,genes])) %>%
        gather(expression, value, -mouse, -sex, -DOwave, -diet_days) %>%
        group_by(expression) %>%
        nest()
mod_fxn = function(df) {
  lm(value ~ sex + DOwave + diet_days, data = df)
}
tmp = tmp %>%
  mutate(model = map(data, mod_fxn)) %>%
  mutate(summ = map(model, tidy)) %>%
  unnest(summ) 
#  kable(tmp, caption = "Effects of Sex, Wave & Diet Days on Expression")

tmp

tmp %>%
  filter(term != "(Intercept)") %>%
  mutate(neg.log.p = -log10(p.value)) %>%
  ggplot(aes(term, neg.log.p)) +
    geom_point() +
    facet_wrap(~expression, ncol=10) +
    labs(title = "Significance of Sex, Wave & Diet Days on Expression") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
rm(tmp)

We can see that `DOwave` is the most significant.  However, given that a few are influenced by `sex` and `diet_days`, we will have to correct for those as well. 
~~~
{: .language-r}



~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
pheno_clin$diet_days = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave + diet_days, data = pheno_clin)
~~~
{: .language-r}

### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!  We are also going to save our qtl results in an `Rdata` file to be used in further lessons. 

### QTL Scans


~~~
qtl.file = "../results/gene.norm_qtl_cis.trans.Rdata"

if(file.exists(qtl.file)) {
  load(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = norm[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    save(qtl, file = qtl.file)
    }
~~~
{: .language-r}

### QTL plots

Let's plot the first 20 gene expression phenotypes.  If you would like to plot all 50, change `for(i in 1:20)` in the code below to `for(i in 1:ncol(qtl))`.




~~~
par(mfrow=c(3,4))
for(i in 1:20) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_plots-1.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-qtl_plots-2.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" />

### QTL Peaks

We are also going to save our peak results so we can use these again else where.  First, lets get out peaks with a LOD score greater than 6. 


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
~~~
{: .language-r}

We will save these peaks into a csv file. 


~~~
kable(peaks %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")

write_csv(peaks, "../results/gene.norm_qtl_peaks_cis.trans.csv")
~~~
{: .language-r}


Table: Phenotype QTL Peaks with LOD >= 6

|lodcolumn          |chr |        pos|       lod|      ci_lo|      ci_hi|
|:------------------|:---|----------:|---------:|----------:|----------:|
|ENSMUSG00000004096 |1   |  12.859898|  6.668630|   5.632546|  21.792678|
|ENSMUSG00000026778 |1   |  72.194159|  6.771007|  71.510439|  73.276997|
|ENSMUSG00000026778 |1   |  91.221683|  9.084761|  89.864283|  92.975712|
|ENSMUSG00000097707 |1   | 183.874168|  6.172552| 127.038745| 184.617316|
|ENSMUSG00000045374 |2   |   4.700297|  6.853078|   4.241948|   4.929193|
|ENSMUSG00000042647 |2   |  24.643031|  7.536545|  22.269496|  57.433658|
|ENSMUSG00000048612 |2   |  26.091626|  6.509792|  25.716296|  26.470332|
|ENSMUSG00000059742 |2   |  62.663310| 78.449795|  62.655686|  63.365553|
|ENSMUSG00000029368 |2   | 105.163834|  7.394643| 103.885974| 106.462173|
|ENSMUSG00000093059 |2   | 123.652615|  6.582997|  93.511738| 125.199675|
|ENSMUSG00000030386 |2   | 125.079199|  6.964568| 124.404221| 125.183892|
|ENSMUSG00000045733 |2   | 125.552992|  6.535442| 109.256226| 150.811972|
|ENSMUSG00000026778 |2   | 133.398934|  6.712826|  76.339229| 144.786441|
|ENSMUSG00000078868 |2   | 174.924525| 23.280768| 174.875030| 177.473149|
|ENSMUSG00000054659 |4   |  33.245486| 20.772250|  32.884827|  33.275701|
|ENSMUSG00000028411 |4   |  40.670078| 11.746247|  39.098259|  41.049715|
|ENSMUSG00000029368 |4   | 117.436515|  6.220128| 110.453461| 119.756454|
|ENSMUSG00000041120 |4   | 138.690361| 20.620303| 138.608123| 139.946675|
|ENSMUSG00000004096 |4   | 142.050722|  6.578088| 140.597905| 144.510585|
|ENSMUSG00000033597 |5   |  74.773745|  8.114646|  71.873670|  85.331839|
|ENSMUSG00000022370 |5   | 104.062885|  7.557708| 102.945105| 104.949229|
|ENSMUSG00000073016 |5   | 106.496702|  6.704994| 103.718321| 146.694026|
|ENSMUSG00000029591 |5   | 113.915256| 41.615471| 113.384269| 114.268350|
|ENSMUSG00000063430 |5   | 114.268350| 61.555183| 113.913594| 114.268350|
|ENSMUSG00000042647 |5   | 121.702189| 77.003675| 121.660329| 121.713435|
|ENSMUSG00000054659 |6   |   5.435945|  6.766848|   4.141181|  17.558585|
|ENSMUSG00000037845 |6   |  53.508095|  7.033017|  48.248399|  53.529938|
|ENSMUSG00000022336 |6   |  53.655019|  7.986251|  16.936091|  55.958893|
|ENSMUSG00000021668 |6   |  53.877659|  7.650088|  53.567586|  54.059609|
|ENSMUSG00000063018 |7   |   6.399629|  7.289107|   5.092147|  16.163886|
|ENSMUSG00000053889 |7   |  12.721644|  6.005457|   5.840839|  80.805709|
|ENSMUSG00000030386 |7   |  12.997827|  9.343187|   7.421153|  15.910616|
|ENSMUSG00000029591 |7   |  31.065202|  6.016437|  30.444496|  34.984291|
|ENSMUSG00000066513 |7   |  44.364954| 17.217957|  43.746502|  44.482294|
|ENSMUSG00000078868 |7   |  44.940291|  6.410139|  37.782379|  98.521991|
|ENSMUSG00000058444 |7   | 115.002417|  6.054025|  54.468380| 117.692619|
|ENSMUSG00000031384 |7   | 133.917441|  6.042874| 133.166981| 134.040585|
|ENSMUSG00000045733 |7   | 140.177559| 55.621286| 140.126351| 140.177559|
|ENSMUSG00000022370 |8   |  18.865591|  6.092211|  18.445860|  90.886876|
|ENSMUSG00000052616 |8   | 104.578805| 58.862907| 104.243290| 104.578805|
|ENSMUSG00000004096 |9   |  14.652765| 39.633604|  13.855991|  14.703558|
|ENSMUSG00000001349 |9   |  15.801722|  6.164058|   9.877068| 110.630801|
|ENSMUSG00000053889 |9   |  34.995506| 68.533287|  34.813042|  35.013994|
|ENSMUSG00000058444 |9   |  63.267626|  9.664105|  61.282370|  63.997996|
|ENSMUSG00000053889 |9   |  67.713523|  6.347508|  61.984246|  68.890051|
|ENSMUSG00000028411 |9   | 108.200819|  6.379825| 105.842326| 111.736252|
|ENSMUSG00000038152 |9   | 109.538108|  6.281700| 106.579404| 111.547928|
|ENSMUSG00000093059 |10  |   8.209153|  6.383420|   3.854456|   9.499216|
|ENSMUSG00000037845 |10  |  20.958482|  6.385955|  20.809881|  21.983132|
|ENSMUSG00000045733 |10  |  75.695784|  6.116431|  69.943514| 118.013538|
|ENSMUSG00000020260 |10  |  77.132202| 33.556575|  76.670577|  77.633198|
|ENSMUSG00000035311 |10  |  88.509852| 24.880267|  87.902115|  88.586222|
|ENSMUSG00000053889 |11  |   5.561303|  6.991227|   4.059120|  66.256656|
|ENSMUSG00000097707 |11  |  12.241828|  6.125867|  10.411062|  16.905028|
|ENSMUSG00000020321 |11  |  24.350764| 11.045568|  21.569820|  25.802864|
|ENSMUSG00000005198 |11  |  69.796313| 10.094049|  69.064645|  71.782692|
|ENSMUSG00000029368 |11  | 107.725392|  7.111887| 105.561849| 107.935117|
|ENSMUSG00000041120 |11  | 120.669585|  6.269615| 118.702148| 121.519022|
|ENSMUSG00000045733 |13  |  44.529337|  6.437152|  42.542597|  46.132779|
|ENSMUSG00000038152 |13  |  45.979436| 17.115407|  45.705273|  46.976410|
|ENSMUSG00000078868 |13  |  59.014905|  6.579179|  21.948789|  63.537793|
|ENSMUSG00000001349 |13  |  94.002661|  6.317148|  92.209903| 101.137104|
|ENSMUSG00000021668 |13  |  97.096191| 10.159156|  95.927238|  97.902562|
|ENSMUSG00000087002 |13  | 118.830394|  6.419148| 118.229414| 120.387272|
|ENSMUSG00000045733 |14  |  24.728913|  6.056478|  19.636097|  32.308130|
|ENSMUSG00000029591 |14  | 112.832347|  6.214735|  25.415171| 119.449546|
|ENSMUSG00000001349 |15  |  21.199357|  6.524449|  14.694622|  50.177210|
|ENSMUSG00000022336 |15  |  42.502340| 13.305623|  41.164200|  43.790803|
|ENSMUSG00000022370 |15  |  55.188785| 22.999712|  54.838771|  56.008582|
|ENSMUSG00000042647 |15  |  77.634457|  6.123942|  76.112162|  78.043784|
|ENSMUSG00000020580 |16  |   9.982367|  6.811876|   5.324393|  77.459783|
|ENSMUSG00000005198 |16  |  10.471652|  7.600821|   7.267803|  12.579260|
|ENSMUSG00000033177 |16  |  45.712180| 50.417700|  45.680294|  45.715820|
|ENSMUSG00000033177 |16  |  48.827759| 42.433845|  48.524425|  49.265132|
|ENSMUSG00000033177 |16  |  51.305847| 31.463734|  51.101502|  51.610310|
|ENSMUSG00000033597 |17  |  24.311984| 15.498613|  23.951943|  25.404500|
|ENSMUSG00000042647 |17  |  54.316877|  6.716774|  45.007252|  56.032945|
|ENSMUSG00000093514 |17  |  68.475585|  6.315485|  67.760641|  69.556918|
|ENSMUSG00000093751 |18  |  35.972927| 11.157035|  31.941452|  38.254377|
|ENSMUSG00000022336 |18  |  36.747599|  6.093247|  36.074158|  38.270815|
|ENSMUSG00000097707 |18  |  55.068951|  6.676204|  25.764916|  57.257690|
|ENSMUSG00000072692 |18  |  58.121373|  6.245350|  29.978305|  63.207014|
|ENSMUSG00000026778 |18  |  65.035321|  6.283797|  57.395640|  67.917363|
|ENSMUSG00000093751 |18  |  78.181640|  8.106581|  77.308149|  80.650891|
|ENSMUSG00000048612 |19  |  38.051674| 48.154394|  37.904186|  38.072015|
|ENSMUSG00000035311 |X   |  49.934828|  8.992905|  46.287980|  51.029010|
|ENSMUSG00000031384 |X   |  52.691279|  6.483433|  41.456110|  84.889868|
|ENSMUSG00000073016 |X   | 104.475947| 28.222084| 104.468740| 105.509305|
|ENSMUSG00000097707 |X   | 106.883167|  7.350023| 101.385909| 107.440271|
|ENSMUSG00000031256 |X   | 134.062872| 12.450128| 131.658191| 134.163228|
|ENSMUSG00000031256 |X   | 158.717231|  8.237609| 152.567174| 160.846608|
|ENSMUSG00000031384 |X   | 166.829809| 17.049244| 164.552669| 167.134378|

### QTL Peaks Figure


~~~
qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_heatmap-1.png" alt="plot of chunk qtl_heatmap" width="612" style="display: block; margin: auto;" />

> ## Challenge
> What do the qtl scans for all gene exression traits look like? *Note:* Don't worry, we've done the qtl scans for you!!!
> You can read in this file, `../data/gene.norm_qtl_all.genes.Rdata`, which are the `scan1` results for all gene expression traits. 
>
> > ## Solution
> > 
> > 
> > ~~~
> > load("../data/gene.norm_qtl_all.genes.Rdata")
> > 
> > lod_threshold = 6
> > peaks = find_peaks(scan1_output = qtl, 
> >                map = map, 
> >                threshold = lod_threshold, 
> >                peakdrop = 4, 
> >                prob = 0.95)
> > write_csv(peaks, "../results/gene.norm_qtl_all.genes_peaks.csv")
> > 
> > ## Heat Map
> > qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
> > ~~~
> > {: .language-r}
> {: .solution}
{: .challenge}
