---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 06-map-many-eqtls.md in _episodes_rmd/
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "To map several genes at the same time"
source: Rmd
---



### Load Libraries  

For this lesson, we need to install two more libraries and load them.  You can do this by typing the following code:


~~~
BiocManager::install(c("AnnotationHub","rtracklayer"))
~~~
{: .language-r}

Let's install our libraries, and source two other R scripts.


~~~
library(tidyverse)
library(knitr)
library(broom)
library(qtl2)
library(qtl2ggplot)
library(RColorBrewer)

source("../code/gg_transcriptome_map.R")
source("../code/qtl_heatmap.R")
~~~
{: .language-r}

Before we begin this lesson, we need to create another directory called `results` in our main directory.  You can do this by clicking on the "Files" tab and navigate into the main directory.  Then select "New Folder" and name it "results".

### Load Data


~~~
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# data from paper
load("../data/dataset.islet.rnaseq.RData")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

# genotype probabilities
probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")
~~~
{: .language-r}

### Data Selection

For this lesson, lets choose a random set of 50 gene expression phenotypes.




~~~
genes = colnames(norm)

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[sams]

gene.info <- dataset.islet.rnaseq$annots[genes,]
rownames(gene.info) = NULL
kable(gene.info[1:10,])
~~~
{: .language-r}



|gene_id            |symbol        |chr |      start|        end| strand|     middle|nearest.marker.id |biotype        |module         |hotspot |
|:------------------|:-------------|:---|----------:|----------:|------:|----------:|:-----------------|:--------------|:--------------|:-------|
|ENSMUSG00000030339 |Ltbr          |6   | 125.306571| 125.313885|     -1| 125.310228|6_125336952       |protein_coding |royalblue      |NA      |
|ENSMUSG00000028024 |Enpep         |3   | 129.269175| 129.332720|     -1| 129.300948|3_129312957       |protein_coding |grey           |NA      |
|ENSMUSG00000040693 |Slco4c1       |1   |  96.818784|  96.872171|     -1|  96.845478|1_96845353        |protein_coding |grey           |NA      |
|ENSMUSG00000028150 |Rorc          |3   |  94.372794|  94.398276|      1|  94.385535|3_94387832        |protein_coding |grey           |NA      |
|ENSMUSG00000025092 |Hspa12a       |19  |  58.795751|  58.860984|     -1|  58.828368|19_58753833       |protein_coding |white          |NA      |
|ENSMUSG00000097537 |2610020C07Rik |16  |  11.203383|  11.225796|      1|  11.214590|16_11193504       |lincRNA        |darkolivegreen |NA      |
|ENSMUSG00000074059 |Fbxw18        |9   | 109.676734| 109.702700|     -1| 109.689717|9_109686322       |protein_coding |grey           |NA      |
|ENSMUSG00000079469 |Pigb          |9   |  73.007419|  73.040378|     -1|  73.023898|9_73024958        |protein_coding |grey60         |NA      |
|ENSMUSG00000039714 |Cplx3         |9   |  57.600019|  57.606281|     -1|  57.603150|9_57602989        |protein_coding |grey           |NA      |
|ENSMUSG00000096667 |Gm22862       |16  |   3.353752|   3.353872|     -1|   3.353812|16_3355414        |miRNA          |grey           |NA      |

### Expression Data

Lets check the distribution for the first 20 gene expression phenotypes. If you would like to check the distribution of all 50 genes, change `for(gene in genes[1:20])` in the code below to `for(gene in genes)`.


~~~
par(mfrow=c(3,4))
for(gene in genes[1:20]){
  hist(norm[,gene], main = gene)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-hist_untransformed-2.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

Check the distributions.  Do they all have a normal distribution?

You will notice that the distribtion of some genes are skewed to the left.  This means that that only a small amount of samples have data and therefore, will need to be removed.  A suitable qc would be keeping expression data that have at least 5% of the samples with more than 10 reads.


~~~
genes_qc <- which(as.numeric(colSums(counts[,genes] > 10)) >= 0.05 * nrow(counts[,genes]))
genes <- genes[genes_qc]
~~~
{: .language-r}

### The Marker Map  

We are using the same marker map as in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#the-marker-map)


### Genotype probabilities  

We have explored this earlier in th previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#genotype-probabilities).  But, as a reminder, we have already calculated genotype probabilities which we loaded above called `probs`.  This contains the 8 state genotype probabilities using the 69k grid  map of the same 500 DO mice that also have clinical phenotypes. 


### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

We have explored the kinship matrix in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#kinship-matrix). It has already been calculated and loaded in above. 


### Covariates    

Now let's add the necessary covariates. For these 50 gene expression data, we will correct for `DOwave`,`sex` and `diet_days`.





~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
pheno_clin$diet_days = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave + diet_days, data = pheno_clin)
~~~
{: .language-r}

### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!  We are also going to save our qtl results in an `Rdata` file to be used in further lessons. 

### QTL Scans


~~~
qtl.file = "../results/gene.norm_qtl_cis.trans.Rdata"

if(file.exists(qtl.file)) {
  load(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = norm[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    save(qtl, file = qtl.file)
    }
~~~
{: .language-r}

### QTL plots

Let's plot the first 20 gene expression phenotypes.  If you would like to plot all 50, change `for(i in 1:20)` in the code below to `for(i in 1:ncol(qtl))`.




~~~
par(mfrow=c(3,4))
for(i in 1:20) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_plots-1.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-qtl_plots-2.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" />

### QTL Peaks

We are also going to save our peak results so we can use these again else where.  First, lets get out peaks with a LOD score greater than 6. 


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
~~~
{: .language-r}

We will save these peaks into a csv file. 


~~~
kable(peaks[1:10,] %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")

write_csv(peaks, "../results/gene.norm_qtl_peaks_cis.trans.csv")
~~~
{: .language-r}


Table: Phenotype QTL Peaks with LOD >= 6

|lodcolumn          |chr |       pos|       lod|      ci_lo|     ci_hi|
|:------------------|:---|---------:|---------:|----------:|---------:|
|ENSMUSG00000030339 |2   | 165.48591|  6.209535|   8.690542| 165.72050|
|ENSMUSG00000028024 |3   | 129.39067| 92.464341| 129.382808| 129.47164|
|ENSMUSG00000025092 |5   | 146.68827|  6.233493|   9.852555| 151.83362|
|ENSMUSG00000030339 |6   | 125.33695| 56.844968| 125.226964| 125.47192|
|ENSMUSG00000030339 |7   |  30.52771|  7.135398|  29.686889|  34.42942|
|ENSMUSG00000028024 |7   | 109.09633|  8.095317| 103.288648| 109.27894|
|ENSMUSG00000097537 |12  | 117.98840|  6.851064| 116.974351| 120.12062|
|ENSMUSG00000025092 |14  |  88.57777|  6.539342|  80.478741|  92.35269|
|ENSMUSG00000097537 |16  |  10.47957| 13.931060|   9.971609|  11.87123|
|ENSMUSG00000025092 |19  |  58.91065| 26.056973|  58.686130|  59.01886|

### QTL Peaks Figure


~~~
qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_heatmap-1.png" alt="plot of chunk qtl_heatmap" width="612" style="display: block; margin: auto;" />

> ## Challenge
> What do the qtl scans for all gene exression traits look like? *Note:* Don't worry, we've done the qtl scans for you!!!
> You can read in this file, `../data/gene.norm_qtl_all.genes.Rdata`, which are the `scan1` results for all gene expression traits. 
>
> > ## Solution
> > 
> > 
> > ~~~
> > load("../data/gene.norm_qtl_all.genes.Rdata")
> > 
> > lod_threshold = 6
> > peaks = find_peaks(scan1_output = qtl.all, 
> >                map = map, 
> >                threshold = lod_threshold, 
> >                peakdrop = 4, 
> >                prob = 0.95)
> > write_csv(peaks, "../results/gene.norm_qtl_all.genes_peaks.csv")
> > 
> > ## Heat Map
> > qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
> > ~~~
> > {: .language-r}
> {: .solution}
{: .challenge}
