---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 06-map-many-eqtls.md in _episodes_rmd/
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "To map several genes at the same time"
source: Rmd
---



### Load Libraries  

For this lesson, we need to install two more libraries and load them.  You can do this by typing the following code:


~~~
BiocManager::install(c("AnnotationHub","rtracklayer"))
~~~
{: .language-r}

Let's install our libraries, and source two other R scripts.


~~~
library(tidyverse)
library(knitr)
library(broom)
library(qtl2)
library(qtl2ggplot)
library(RColorBrewer)

source("../code/gg_transcriptome_map.R")
~~~
{: .language-r}



~~~
Error in library(AnnotationHub): there is no package called 'AnnotationHub'
~~~
{: .error}



~~~
source("../code/qtl_heatmap.R")
~~~
{: .language-r}

Before we begin this lesson, we need to create another directory called `results` in our main directory.  You can do this by clicking on the "Files" tab and navigate into the main directory.  Then select "New Folder" and name it "results".

### Load Data


~~~
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# data from paper
load("../data/dataset.islet.rnaseq.RData")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

# genotype probabilities
probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")
~~~
{: .language-r}

### Data Selection

For this lesson, lets choose a random set of 50 gene expression phenotypes.




~~~
genes = colnames(norm)

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[sams]

gene.info <- dataset.islet.rnaseq$annots[genes,]
rownames(gene.info) = NULL
kable(gene.info[1:10,])
~~~
{: .language-r}



|gene_id            |symbol   |chr |      start|        end| strand|     middle|nearest.marker.id |biotype        |module  |hotspot |
|:------------------|:--------|:---|----------:|----------:|------:|----------:|:-----------------|:--------------|:-------|:-------|
|ENSMUSG00000003282 |Plag1    |4   |   3.900996|   3.938423|     -1|   3.919710|4_3922207         |protein_coding |white   |NA      |
|ENSMUSG00000020184 |Mdm2     |10  | 117.688888| 117.710758|     -1| 117.699823|10_117696312      |protein_coding |orange  |NA      |
|ENSMUSG00000058952 |Cfi      |3   | 129.836739| 129.875328|      1| 129.856034|3_129918506       |protein_coding |grey    |NA      |
|ENSMUSG00000052305 |Hbb-bs   |7   | 103.826534| 103.828096|     -1| 103.827315|7_103735454       |protein_coding |grey    |NA      |
|ENSMUSG00000055733 |Nap1l3   |X   | 122.394565| 122.397385|     -1| 122.395975|X_122228767       |protein_coding |green   |NA      |
|ENSMUSG00000068606 |Gm4841   |18  |  60.268301|  60.273267|     -1|  60.270784|18_60231339       |protein_coding |magenta |NA      |
|ENSMUSG00000068480 |Gm7551   |7   |  65.785141|  65.785932|      1|  65.785536|7_65765622        |pseudogene     |grey    |NA      |
|ENSMUSG00000098104 |Gm6085   |1   |   4.687934|   4.689403|     -1|   4.688668|1_4548703         |pseudogene     |grey    |NA      |
|ENSMUSG00000021264 |Yy1      |12  | 108.792973| 108.816632|      1| 108.804802|12_108782354      |protein_coding |brown   |NA      |
|ENSMUSG00000032875 |Arhgef17 |7   | 100.869746| 100.932161|     -1| 100.900954|7_100914347       |protein_coding |purple  |NA      |

### Expression Data

Lets check the distribution for the first 20 gene expression phenotypes. If you would like to check the distribution of all 50 genes, change `for(gene in genes[1:20])` in the code below to `for(gene in genes)`.


~~~
par(mfrow=c(3,4))
for(gene in genes[1:20]){
  hist(norm[,gene], main = gene)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-hist_untransformed-2.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

Check the distributions.  Do they all have a normal distribution?

You will notice that the distribtion of some genes are skewed to the left.  This means that that only a small amount of samples have data and therefore, will need to be removed.  A suitable qc would be keeping expression data that have at least 5% of the samples with more than 10 reads.


~~~
genes_qc <- which(as.numeric(colSums(counts[,genes] > 10)) >= 0.05 * nrow(counts[,genes]))
genes <- genes[genes_qc]
~~~
{: .language-r}

### The Marker Map  

We are using the same marker map as in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#the-marker-map)


### Genotype probabilities  

We have explored this earlier in th previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#genotype-probabilities).  But, as a reminder, we have already calculated genotype probabilities which we loaded above called `probs`.  This contains the 8 state genotype probabilities using the 69k grid  map of the same 500 DO mice that also have clinical phenotypes. 


### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

We have explored the kinship matrix in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#kinship-matrix). It has already been calculated and loaded in above. 


### Covariates    

Now let's add the necessary covariates. For these 50 gene expression data, we will correct for `DOwave`,`sex` and `diet_days`.





~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
pheno_clin$diet_days = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave + diet_days, data = pheno_clin)
~~~
{: .language-r}

### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!  We are also going to save our qtl results in an `Rdata` file to be used in further lessons. 

### QTL Scans


~~~
qtl.file = "../results/gene.norm_qtl_cis.trans.Rdata"

if(file.exists(qtl.file)) {
  load(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = norm[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    save(qtl, file = qtl.file)
    }
~~~
{: .language-r}



~~~
Warning in gzfile(file, "wb"): cannot open compressed file '../results/
gene.norm_qtl_cis.trans.Rdata', probable reason 'Invalid argument'
~~~
{: .warning}



~~~
Error in gzfile(file, "wb"): cannot open the connection
~~~
{: .error}

### QTL plots

Let's plot the first 20 gene expression phenotypes.  If you would like to plot all 50, change `for(i in 1:20)` in the code below to `for(i in 1:ncol(qtl))`.




~~~
par(mfrow=c(3,4))
for(i in 1:20) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_plots-1.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-qtl_plots-2.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" />

### QTL Peaks

We are also going to save our peak results so we can use these again else where.  First, lets get out peaks with a LOD score greater than 6. 


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
~~~
{: .language-r}

We will save these peaks into a csv file. 


~~~
kable(peaks[1:10,] %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")

write_csv(peaks, "../results/gene.norm_qtl_peaks_cis.trans.csv")
~~~
{: .language-r}


Table: Phenotype QTL Peaks with LOD >= 6

|lodcolumn          |chr |       pos|       lod|     ci_lo|     ci_hi|
|:------------------|:---|---------:|---------:|---------:|---------:|
|ENSMUSG00000058952 |3   | 129.72847| 72.536897| 129.64361| 129.91851|
|ENSMUSG00000003282 |5   | 137.00639|  6.065932|  19.20559| 138.73485|
|ENSMUSG00000055733 |6   |  55.54583|  7.013031|  53.51902| 124.49064|
|ENSMUSG00000003282 |7   |  45.62219|  8.153162|  44.72134|  46.36880|
|ENSMUSG00000052305 |7   |  79.24486|  6.841924|  76.03982| 109.09397|
|ENSMUSG00000020184 |8   |  34.68507|  6.193078|  15.85660|  84.11561|
|ENSMUSG00000020184 |10  |  99.03240|  7.394502|  96.42540| 100.55667|
|ENSMUSG00000068606 |10  | 115.77734|  6.053313| 109.92949| 122.50375|
|ENSMUSG00000020184 |10  | 117.89731| 10.319738| 116.83728| 117.97266|
|ENSMUSG00000052305 |16  |  92.61709|  6.383724|  91.07674|  94.89728|



~~~
Error: Cannot open file for writing:
* '../results/gene.norm_qtl_peaks_cis.trans.csv'
~~~
{: .error}

### QTL Peaks Figure


~~~
qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_heatmap-1.png" alt="plot of chunk qtl_heatmap" width="612" style="display: block; margin: auto;" />

> ## Challenge
> What do the qtl scans for all gene exression traits look like? *Note:* Don't worry, we've done the qtl scans for you!!!
> You can read in this file, `../data/gene.norm_qtl_all.genes.Rdata`, which are the `scan1` results for all gene expression traits. 
>
> > ## Solution
> > 
> > 
> > ~~~
> > load("../data/gene.norm_qtl_all.genes.Rdata")
> > 
> > lod_threshold = 6
> > peaks = find_peaks(scan1_output = qtl.all, 
> >                map = map, 
> >                threshold = lod_threshold, 
> >                peakdrop = 4, 
> >                prob = 0.95)
> > write_csv(peaks, "../results/gene.norm_qtl_all.genes_peaks.csv")
> > 
> > ## Heat Map
> > qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
> > ~~~
> > {: .language-r}
> {: .solution}
{: .challenge}
