---
title: "Review Mapping Steps"
teaching: 30
exercises: 30
questions:
- "??????"
objectives:
- "????"
keypoints:
- "????"
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("05-")
```

### Load Libraries  

```{r load libraries}
library(tidyverse)
library(qtl2)
library(qtl2convert)
#library(qtl2db)
library(GGally)
library(broom)
library(knitr)
library(corrplot)
library(RColorBrewer)

```

## Load Data

```{r load_dependencies}

##phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

##mapping data
load("../data/attie_DO500_mapping.data.RData")

genoprobs <- readRDS("../data/attie_DO500_genoprobs_qtlviewer_8state_69k.rds")

```

### Phenotypes

Lets choose a phenotype.  Here, we have chosen `weight_6wk` (weight taken at 6 weeks).  


Lets check the distribution


```{r hist_untransformed}

hist(pheno_clin$Ins_tAUC, main = "insulin tAUC")

```

Apply the `log()` function to this data.

```{r log_transform}

pheno_clin$Ins_tAUC_log <- log(pheno_clin$Ins_tAUC)

```

Now, lets make a histogram of the log-transformed data.

```{r hist_log_transform}

hist(pheno_clin$Ins_tAUC_log, main = "insulin tAUC (log-transformed)")

```

This looks much better!

### The Marker Map  

The marker map for each chromosome is stored in the `map` object. This is used to plot the LOD scores calculated at each marker during QTL mapping.  Here we are using the 69K grid marker file

```{r map information, include=FALSE}

map=map

```

### Genotype probabilities  

We have already claculated genotype probabilities which we load above

```{r DOex_calc_genoprob}


probs = genoprobs

```

```{r dim_probs}

dim(probs[[1]])

```

```{r geno_plot, fig.width=8, fig.height=6}

plot_genoprob(probs, map, ind = 1, chr = 1)

```

### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

The kinship matrix has already been calculated and loaded in above


```{r kinship_probs, fig.width=8, fig.height=8}

n_samples <- 50

heatmap(K[[1]][1:n_samples, 1:n_samples])

```

### Covariates    

Now lets add the necessary covariates. For these analysis, lets see which covariates are significant (???)

```{r covariates}

# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave, data = pheno_clin)

```
### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!

```{r QTL, warning=FALSE, eval=FALSE}

Ins_tAUC_qtl = scan1(genoprobs = probs, pheno = pheno_clin[,"Ins_tAUC_log", drop = FALSE], kinship = K, addcovar = covar, cores = 2)
```

```{r saveQTL, warning=FALSE, error=FALSE}
# save out the file so that genoprobs don't have to be up on github
saveRDS(Ins_tAUC_qtl, file = "../data/ins_tAUC_log_scan.rds")
```

Lets plot it


```{r qtl_plot, fig.width=8, fig.height=6, warning=FALSE}
# read file back in again
Ins_tAUC_qtl <- readRDS(file = "../data/ins_tAUC_log_scan.rds")
#pdf("qtl_mapping_insulin.tauc.pdf", width=11, height=8)
plot_scan1(x = Ins_tAUC_qtl, map = map, lodcolumn = "Ins_tAUC_log")
  abline(h = 6, col = 2, lwd = 2)
#dev.off()

```

### [Performing a permutation test](https://smcclatchy.github.io/mapping/10-perform-perm-test/) 

Not in scripts

### [Finding LOD peaks](https://smcclatchy.github.io/mapping/07-find-lod-peaks/)

Lets find LOD peaks

```{r interval}
lod_threshold = 6
peaks = find_peaks(scan1_output = Ins_tAUC_qtl, map = map, threshold = lod_threshold, peakdrop = 4, prob = 0.95)
kable(peaks %>% select (-lodindex) %>% arrange(chr, pos), caption = "Phenotype QTL Peaks with LOD >= 6")

```

Lets plot them:

should we?

### [Estimated QTL effects](https://smcclatchy.github.io/mapping/11-est-qtl-effects/) 

Not in scripts


### [SNP Association Mapping](https://smcclatchy.github.io/mapping/12-snp-assoc/)

Not in scripts

### Searching for Candidate Genes

Not in scripts
