---
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "????"
keypoints:
- "????"
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

### Load Libraries  

```{r load libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(qtl2)
library(qtl2convert)
library(GGally)
library(broom)
library(knitr)
library(corrplot)
library(RColorBrewer)
library(qtl2ggplot)

source("../code/gg_transcriptome_map.R")
source("../code/qtl_heatmap.R")
```

## Load Data

```{r load_dependencies}
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")
```


### Expression Data
Lets check the distribution

```{r hist_untransformed}
hist(counts$ENSMUSG00000020679, main = "Hnf1b (counts)")
hist(norm$ENSMUSG00000020679, main = "Hnf1b (norm)")
```

These counts are normalised

### The Marker Map  

The marker map for each chromosome is stored in the `map` object. This is used 
to plot the LOD scores calculated at each marker during QTL mapping.  Here we 
are using the 69K grid marker file.

```{r map information, include=FALSE}


```

### Genotype probabilities  

We have already calculated genotype probabilities which we load above.

```{r DOex_calc_genoprob}



```

### Covariates    

Now lets add the necessary covariates. For these analysis, lets see which 
covariates are significant (???)

```{r covariates}
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave, data = pheno_clin)
```
### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!

Pick 50 random genes (cis and trans genes)

### QTL Scans

```{r qtl_scans}
genes = colnames(counts)
chr11 = which(genes=="ENSMUSG00000020679")
genes = genes[-chr11]

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[c(chr11, sams)]

qtl.file = "../data/gene.counts_qtl_cis.trans_random.rds"

qtl = NULL

if(file.exists(qtl.file)) {
  qtl = readRDS(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = counts[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    saveRDS(qtl, file = qtl.file)
    }
```

### QTL plots

```{r set_plot_rows_columns, echo=FALSE}
par(mfrow=c(17, 3))
```

```{r qtl_plots}
for(i in 1:ncol(qtl)) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
```

### QTL Peaks

```{r qtl_peaks}
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
```

```{r short_qtl_peaks_table, echo=FALSE}
peaks %>%
  dplyr::select(-lodindex) %>%
  arrange(chr, pos) %>%
  kable(caption = "Expression QTL (eQTL) Peaks with LOD >= 6") %>%
  head(n=12)
```

```{r qtl_peaks_table, eval=FALSE}
kable(peaks %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")

write_csv(peaks, "../data/gene.counts_qtl_peaks_random.csv")
```

### QTL Peaks Figure

```{r qtl_peaks_figure}
peaks = peaks %>%
  arrange(lodcolumn)
plot_peaks(peaks, 
           map, 
           col = c("blue","red"), 
           lwd = 3, 
           tick_height = 0.8, 
           gap = 0, 
           main = "LOD > 6")
box()
ggplot_peaks(peaks, 
             map, 
             col = c("blue","red"), 
             legend.title = "LOD > 6") 
```

```{r qtl_heatmap}

qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)

```

